File: native/macos/clipboard.m
/* https://github.com/takeiteasy/paul

paul Copyright (C) 2025 George Watson

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>. */

#ifndef PAUL_NO_CLIPBOARD
#include "../clipboard.h"
#import <AppKit/AppKit.h>

bool paul_clipboard_available(void) {
    return [NSPasteboard generalPasteboard] != NULL;
}

const char *paul_clipboard_get(void) {
    NSPasteboard *pb = [NSPasteboard generalPasteboard];
    if ([[pb types] containsObject:NSPasteboardTypeString]) {
        NSString *text = [pb stringForType:NSPasteboardTypeString];
        return strdup([text UTF8String]);
    }
    return NULL;
}

void paul_clipboard_set(const char *str) {
    NSPasteboard *pb = [NSPasteboard generalPasteboard];
    [pb clearContents];
    [pb setString:@(str)
          forType:NSPasteboardTypeString];
}
#endif // PAUL_NO_CLIPBOARD

File: native/linux/backends/gtk/clipboard.c
/* https://github.com/takeiteasy/hal

 hal Copyright (C) 2025 George Watson

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <https://www.gnu.org/licenses/>. */

#ifndef HAL_NO_CLIPBOARD
#include "../../../clipboard.h"
#include "../../internal.h"
#include <gtk/gtk.h>

bool hal_clipboard_available(void) {
    return true;
}

const char *hal_clipboard_get(void) {
    GtkClipboard *clipboard = gtk_clipboard_get(GDK_SELECTION_CLIPBOARD);
    gchar *text = gtk_clipboard_wait_for_text(clipboard);
    return text ? strdup((char*)text) : NULL;
}

void hal_clipboard_set(const char *str) {
    GtkClipboard *clipboard = gtk_clipboard_get(GDK_SELECTION_CLIPBOARD);
    gtk_clipboard_set_text(clipboard, str, -1);
}
#endif // HAL_NO_CLIPBOARD
